docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+# Source Generators Cookbook
+
+## Summary
+
+This document aims to be a guide to help the creation of source generators by providing a series of reusable patterns.
+It also aims to set out what types of generators are possible under the current design and what is explictly out of scope.
+
+> **Note**: The design for the source generator proposal is still under review. This document uses only one possible syntax, and
+> it is expected to change without notice as the feature evolves.
+
+## Proposal
+
+It is worth breifly setting out the design goals of generators here. For a more complete design see the [full design document](source-generators.md)
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additive_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
```

Comment:
Think we want to say something to the effect that source generators can access the full Roslyn API which includes additional files. Maybe want to add a line that additional files are particularly useful for generators that translate non-C# code to C# code like XML files, JSON, etc ...